### PR TITLE
CNF-13328: Handle ordered deletion of rendered manifests

### PR DESCRIPTION
**Pull request by sakhoury**
This commit refactors the SiteConfig finalizer to handle ordered deletion of rendered manifests by descending order of the sync-wave.

### CODE REVIEW

**Comment by sakhoury** [comment]
> @sakhoury How does the controller add sync-waves, e.g. what is the order? During deletion, for example, if managedCluster CR is not deleted at first, and then the hive and assisted Installer CRs as an order, the deletion of managedcluster may take 5 minutes. Have you taken that into consideration or experienced the issue? Thanks

@sabbir-47 the controller is not responsible for adding sync-waves, the user is! This is accomplished by annotating the templates with the `metaclusterinstall.openshift.io/sync-wave` annotation. If the user does not specify a sync-wave, the default value is used. The controller creates/updates in ascending order, and deletes in descending order.

**Comment by sabbir-47** [comment]
> > @sakhoury How does the controller add sync-waves, e.g. what is the order? During deletion, for example, if managedCluster CR is not deleted at first, and then the hive and assisted Installer CRs as an order, the deletion of managedcluster may take 5 minutes. Have you taken that into consideration or experienced the issue? Thanks
> 
> @sabbir-47 the controller is not responsible for adding sync-waves, the user is! This is accomplished by annotating the templates with the `metaclusterinstall.openshift.io/sync-wave` annotation. If the user does not specify a sync-wave, the default value is used. The controller creates/updates in ascending order, and deletes in descending order.

So, we are expecting that user must know all the intricacies and ordered mechanism of the CRs for cluster deployment from their side?

**Comment by sakhoury** [comment]
> > > @sakhoury How does the controller add sync-waves, e.g. what is the order? During deletion, for example, if managedCluster CR is not deleted at first, and then the hive and assisted Installer CRs as an order, the deletion of managedcluster may take 5 minutes. Have you taken that into consideration or experienced the issue? Thanks
> > 
> > 
> > @sabbir-47 the controller is not responsible for adding sync-waves, the user is! This is accomplished by annotating the templates with the `metaclusterinstall.openshift.io/sync-wave` annotation. If the user does not specify a sync-wave, the default value is used. The controller creates/updates in ascending order, and deletes in descending order.
> 
> So, we are expecting that user must know all the intricacies and ordered mechanism of the CRs for cluster deployment from their side?

The default set of assisted installer and IBI templates are annotated. This is more for custom templates!

File: api/v1alpha1/siteconfig_types.go [hunk 1]
```diff
@@ -329,8 +329,12 @@ type ManifestReference struct {
 	APIGroup *string `json:"apiGroup,omitempty"`
 	// Kind is the type of resource being referenced
 	Kind string `json:"kind,omitempty"`
-	// Name is the name of resource being referenced
+	// Name is the name of the resource being referenced
 	Name string `json:"name,omitempty"`
+	// Namespace is the namespace of the resource being referenced
+	Namespace string `json:"namespace,omitempty"`
```
Comment (carbonin) [inline api/v1alpha1/siteconfig_types.go:335]: Not sure this belongs in the sync wave PR.

**Review by carbonin — COMMENTED** [review]
Outside of the extra bits regarding the manifest namespace this looks fine.

File: api/v1alpha1/siteconfig_types.go (hunk 1 above)
Comment (sakhoury) [reply api/v1alpha1/siteconfig_types.go:335]: The namespace is required for specifying the object to be deleted in the finalizer section.

Comment (sakhoury) [reply api/v1alpha1/siteconfig_types.go:335]: The namespace here serves a different purpose. It is used for specifying the namespace of the rendered manifests whereas #7 is for specifying extra-manifests outside of the local namespace. Furthermore, an empty namespace allows us to identify that this is a cluster scoped resource (such as ManagedCluster kind).

Comment (carbonin) [reply api/v1alpha1/siteconfig_types.go:335]: Ah okay, so will this only ever be the namespace of the SiteConfig or empty?

**Review by carbonin — APPROVED** [review]

**Pull request merged**